File: qqbot-module-host/src/main/java/com/mieai/qqbot/module/host/ModuleGraph.java
package com.mieai.qqbot.module.host;

import com.mieai.qqbot.module.api.FrameworkVersion;
import com.mieai.qqbot.module.api.ModuleDependency;
import com.mieai.qqbot.module.api.ModuleDescriptor;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

final class ModuleGraph {
    private ModuleGraph() {}

    static List<String> resolve(Map<String, ModuleDescriptor> descriptors) {
        SemanticVersion frameworkVersion = SemanticVersion.parse(FrameworkVersion.CURRENT);
        for (ModuleDescriptor descriptor : descriptors.values()) {
            descriptor.webContributions().forEach(contribution -> {
                if (contribution.componentKey() != null) {
                    throw new IllegalStateException("External module " + descriptor.id()
                            + " cannot contribute a built-in Web component");
                }
                String expected = "/modules/" + descriptor.id() + "/" + contribution.id();
                if (!expected.equals(contribution.route())) {
                    throw new IllegalStateException("Web contribution " + descriptor.id() + ":"
                            + contribution.id() + " must use route " + expected);
                }
            });
            if (frameworkVersion.compareTo(
                    SemanticVersion.parse(descriptor.minimumFrameworkVersion())) < 0) {
                throw new IllegalStateException("Module " + descriptor.id()
                        + " requires framework " + descriptor.minimumFrameworkVersion());
            }
            for (ModuleDependency dependency : descriptor.dependencies()) {
                ModuleDescriptor target = descriptors.get(dependency.moduleId());
                if (target == null) {
                    if (!dependency.optional()) {
                        throw new IllegalStateException("Module " + descriptor.id()
                                + " requires missing module " + dependency.moduleId());
                    }
                    continue;
                }
                if (SemanticVersion.parse(target.version()).compareTo(
                        SemanticVersion.parse(dependency.minimumVersion())) < 0) {
                    throw new IllegalStateException("Module " + descriptor.id() + " requires "
                            + dependency.moduleId() + " " + dependency.minimumVersion()
                            + " or newer");
                }
            }
        }

        List<String> order = new ArrayList<>();
        Map<String, VisitState> visits = new HashMap<>();
        descriptors.keySet().stream().sorted()
                .forEach(id -> visit(id, descriptors, visits, order, new ArrayList<>()));
        return List.copyOf(order);
    }

    private static void visit(
            String id,
            Map<String, ModuleDescriptor> descriptors,
            Map<String, VisitState> visits,
            List<String> order,
            List<String> path) {
        VisitState state = visits.get(id);
        if (state == VisitState.COMPLETE) return;
        if (state == VisitState.VISITING) {
            path.add(id);
            throw new IllegalStateException("Framework module dependency cycle: "
                    + String.join(" -> ", path));
        }
        visits.put(id, VisitState.VISITING);
        path.add(id);
        descriptors.get(id).dependencies().stream()
                .filter(dependency -> descriptors.containsKey(dependency.moduleId()))
                .sorted(java.util.Comparator.comparing(ModuleDependency::moduleId))
                .forEach(dependency -> visit(dependency.moduleId(), descriptors, visits, order,
                        new ArrayList<>(path)));
        visits.put(id, VisitState.COMPLETE);
        order.add(id);
    }

    private enum VisitState { VISITING, COMPLETE }
}

File: qqbot-module-host/src/main/kotlin/com/mieai/qqbot/module/host/ModuleGraph.kt
package com.mieai.qqbot.module.host

import com.mieai.qqbot.module.api.FrameworkVersion
import com.mieai.qqbot.module.api.ModuleDependency
import com.mieai.qqbot.module.api.ModuleDescriptor

internal object ModuleGraph {
    fun resolve(descriptors: Map<String, ModuleDescriptor>): List<String> {
        val frameworkVersion = SemanticVersion.parse(FrameworkVersion.CURRENT)
        descriptors.values.forEach { descriptor ->
            descriptor.webContributions.forEach { contribution ->
                check(contribution.componentKey == null) {
                    "External module ${descriptor.id} cannot contribute a built-in Web component"
                }
                val expected = "/modules/${descriptor.id}/${contribution.id}"
                check(expected == contribution.route) {
                    "Web contribution ${descriptor.id}:${contribution.id} must use route $expected"
                }
            }
            check(frameworkVersion >= SemanticVersion.parse(descriptor.minimumFrameworkVersion)) {
                "Module ${descriptor.id} requires framework ${descriptor.minimumFrameworkVersion}"
            }
            descriptor.dependencies.forEach dependencyLoop@{ dependency ->
                val target = descriptors[dependency.moduleId]
                if (target == null) {
                    check(dependency.optional) {
                        "Module ${descriptor.id} requires missing module ${dependency.moduleId}"
                    }
                    return@dependencyLoop
                }
                check(SemanticVersion.parse(target.version) >= SemanticVersion.parse(dependency.minimumVersion)) {
                    "Module ${descriptor.id} requires ${dependency.moduleId} ${dependency.minimumVersion} or newer"
                }
            }
        }
        val order = mutableListOf<String>()
        val visits = mutableMapOf<String, VisitState>()
        descriptors.keys.sorted().forEach { visit(it, descriptors, visits, order, mutableListOf()) }
        return order.toList()
    }

    private fun visit(id: String, descriptors: Map<String, ModuleDescriptor>, visits: MutableMap<String, VisitState>, order: MutableList<String>, path: MutableList<String>) {
        when (visits[id]) {
            VisitState.COMPLETE -> return
            VisitState.VISITING -> {
                path.add(id)
                throw IllegalStateException("Framework module dependency cycle: ${path.joinToString(" -> ")}")
            }
            null -> Unit
        }
        visits[id] = VisitState.VISITING
        path.add(id)
        descriptors.getValue(id).dependencies
            .filter { descriptors.containsKey(it.moduleId) }
            .sortedBy { it.moduleId }
            .forEach { visit(it.moduleId, descriptors, visits, order, path.toMutableList()) }
        visits[id] = VisitState.COMPLETE
        order.add(id)
    }

    private enum class VisitState { VISITING, COMPLETE }
}
